File: split/finmlist.pas
function finmlist(p: halfword): halfword;
var
  q: halfword;
begin
  if curlist.auxfield.int <> 0 then{1185:}

  begin
    mem[curlist.auxfield.int + 3].hh.rh := 3;
    mem[curlist.auxfield.int + 3].hh.lh := mem[curlist.headfield].hh.rh;
    if p = 0 then
      q := curlist.auxfield.int
    else
    begin
      q := mem[curlist.auxfield.int + 2].hh.lh;
      if mem[q].hh.b0 <> 30 then
        confusion(876);
      mem[curlist.auxfield.int + 2].hh.lh := mem[q].hh.rh;
      mem[q].hh.rh := curlist.auxfield.int;
      mem[curlist.auxfield.int].hh.rh := p;
    end;
  end{:1185}

  else
  begin
    mem[curlist.tailfield].hh.rh := p;
    q := mem[curlist.headfield].hh.rh;
  end;
  popnest;
  finmlist := q;
end;

File: cpp/finmlist.cpp
#include "finmlist.h"
#include "confusion.h"
#include "popnest.h"

halfword finmlist(halfword p)
{
	halfword q;
	if (curlist.auxfield.int_)
	{
		link(curlist.auxfield.int_+3) = 3;
		info(curlist.auxfield.int_+3) = link(curlist.headfield);
		if (p == 0)
			q = curlist.auxfield.int_;
		else
		{
			q = info(curlist.auxfield.int_+2);
			if (type(q) != 30)
				confusion(876);
			info(curlist.auxfield.int_+2) = link(q);
			link(q) = curlist.auxfield.int_;
			link(curlist.auxfield.int_) = p;
		}
	}
	else
	{
		link(curlist.tailfield) = p;
		q = link(curlist.headfield);
	}
	popnest();
	return q;
}
